Sources/Keychain.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  Copyright © 2020 Rosberry. All rights reserved.
+//
+
+import Foundation
+
+final class Keychain {
+
+    enum KeychainError: Error {
+        case noPassword
+        case unexpectedPasswordData
+        case unexpectedItemData
+        case unhandledError(status: OSStatus)
+        case wrongEncoding
+    }
+
+    let service: String
+
+    init(service: String) {
+        self.service = service
+    }
+
+    func readPassword(forKey key: String) throws -> String {
+        var query = makeKeychainQuery(account: key)
+        query[kSecMatchLimit as String] = kSecMatchLimitOne
+        query[kSecReturnAttributes as String] = kCFBooleanTrue
+        query[kSecReturnData as String] = kCFBooleanTrue
+
+        var queryResult: AnyObject?
+        let status = withUnsafeMutablePointer(to: &queryResult) { pointer in
+            SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer(pointer))
+        }
+
+        guard status != errSecItemNotFound else {
+            throw KeychainError.noPassword
+        }
+        guard status == noErr else {
+            throw KeychainError.unhandledError(status: status)
+        }
+        guard let existingItem = queryResult as? [String: AnyObject],
+            let passwordData = existingItem[kSecValueData as String] as? Data,
+            let password = String(data: passwordData, encoding: .utf8)
+        else {
+            throw KeychainError.unexpectedPasswordData
+        }
+
+        return password
+    }
+
+    func save(value: String?, forKey key: String) throws {
+        guard let value = value else {
+            let query = makeKeychainQuery(account: key)
+            let status = SecItemDelete(query as CFDictionary)
+            guard status == noErr else {
+                throw KeychainError.unhandledError(status: status)
+            }
+            return
+        }
+        guard let encodedPassword = value.data(using: .utf8) else {
+            throw KeychainError.wrongEncoding
+        }
+
+        do {
+            try _ = readPassword(forKey: key)
```

Comment:
Why you call this method here?

Reply:
Same question. Will it throw an `unexpectedPasswordData ` error if keychain has no data for this key?

Reply:
I called it to catch `KeychainError.noPassword` error below. In this case, I know there are no values for passed key.

Reply:
It does not look pretty, but if it is the only way to separate saving and updating logic - then ok